### PR TITLE
Fixing #869. We were trying to iterate over projects that do not support VS Automation (including unloaded).

### PR DESCRIPTION
**Pull request by amanda-tarafa**
This throws an exception and since the commands have dynamic visibility set they stayed grayed out.

### CODE REVIEW

File: GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/ProjectHelper.cs [hunk 1]
```diff
@@ -130,14 +130,18 @@ public ProjectSourceFile FindSourceFile(string sourceFilePath)
...
-                return false;
-            }
+            return project != null
+                // The project is not the miscellaneous project which is created automatically
+                // by VS to include elements not belonging to any other project.
+                && project.Kind != Constants.vsProjectKindMisc
+                // The project does not represent a folder.
+                && project.Kind != Constants.vsProjectKindSolutionItems
+                // The project is not an unmolded project. Unmolded projects do not support
+                // automation. // Several Project properties that are used acros the extension are not available
```
Comment (iantalarico) [inline GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/ProjectHelper.cs]: nit: remove extra "//"

File: GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs [hunk 2]
```diff
@@ -247,8 +247,8 @@ private IParsedProject GetStartupProject()
                 return null;
             }
 
-            string projectName = Path.GetFileNameWithoutExtension(startupProjectFilePath);
-            foreach (Project p in _solution.Projects)
+            string projectName = Path.GetFileNameWithoutExtension(startupProjectFilePath);
```
Comment (iantalarico) [inline GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs]: nit: remove extra whitespace

File: GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs [hunk 3]
```diff
@@ -264,12 +264,9 @@ private IParsedProject GetStartupProject()
         private List<ProjectHelper> GetProjectList()
         {
             List<ProjectHelper> list = new List<ProjectHelper>();
-            foreach (Project project in _solution.Projects)
+            foreach (Project project in Projects)
```
Comment (iantalarico) [inline GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs]: Won't this call itself?

File: GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/ProjectHelper.cs (hunk 1 above)
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/ProjectHelper.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs (hunk 2 above)
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs (hunk 3 above)
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs]: Ohhhh, something I accidentally changed and thought I had changed back, but didn't! Sorry!
Interesting though how running the extension works (a lot of it) and all the tests passed with that. Let me see what I can do about the tests for this.

Comment (iantalarico) [reply GoogleCloudExtension/GoogleCloudExtension/SolutionUtils/SolutionHelper.cs]: Sounds good, just ping me when it's ready!

**Comment by amanda-tarafa** [comment]
Changed PR to the new branch.

**Comment by iantalarico** [comment]
@amanda-tarafa Is this ready to be reviewed?

**Comment by amanda-tarafa** [comment]
@iantalarico will be 10 minutes after the meeting is over, just need to push my changes and add a couple of comments to the PR.

**Comment by amanda-tarafa** [comment]
To fix #869 had to change slightly what projects we consider valid and check for their validity before they were scanned for the contextual menu.
Also, refactor some to "merge" ProjectHelper and IParsedProject since those are two different wrappers on top of the Automation Model Project interface.
Added some unit tests for ProjectHelper.
Some more improvements can be made, but those were too much for this PR.

1. Properly merge ProjectHelper and IParsedProject, most of their properties are the same. That would need though to add unit tests to clients of these class and interface before refactoring.
2. Refactor the code on ProjectHelper to be able to inject System.IO.Path and redo the tests so we don't have that dependency on ProjectHelper tests (this is not a file system dependency, it's just a path manipulation dependency).

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+namespace GoogleCloudExtensionUnitTests.SolutionUtils
+{
+    [TestClass]
+    public class ProjectHelperUnitTest
+    {
+        private const string WebApplicationProjectKind = "{349C5851-65DF-11DA-9384-00065B846F21}";
+
+        private static readonly string ProjectFullName = Path.Combine("Projects", "Solution1", "Project1", "project1.csproj");
+        private static readonly string ProjectRoot = Path.Combine("Projects", "Solution1", "Project1");
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: What do you mean?

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+        private const string UniqueNameNoSeparator = @"project1.csproj";
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: And same here, what do you mean?

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (iantalarico) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: The project name is just he root with "project1.csproj" added at the end.  In the code I posted I just used the root to create the full name.

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iantalarico) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: We can do:

```
private const string UniqueNameNoSeparator = @"project1.csproj";
 private static readonly string ProjectFullName = Path.Combine("Projects", "Solution1", "Project1", UniqueNameNoSeparator);
```

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: Oh, got it, yes!

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        private void InitProjectMockForKindValidityTesting()
+        {
+            _projectMock.Setup(p => p.FullName).Returns(ProjectFullName);
+            _projectMock.Setup(p => p.Properties).Returns(Mock.Of<Properties>());
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException), AllowDerivedTypes = true)]
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        public void TestConstructorProjectKindUnmolded()
+        {
+            InitProjectMockForKindValidityTesting();
+            _projectMock.Setup(p => p.Kind).Returns(Constants.vsProjectKindUnmodeled);
+
+            _objectUnderTest = new ProjectHelper(_projectMock.Object);
+        }
+
+        private void InitProjectMockForExtraValidityTesting()
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        private static readonly string UniqueNameSeparator = Path.DirectorySeparatorChar + @"project1.csproj";
+        private const string UniqueNameDifferent = @"project2.csproj";
+
+
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: done

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+            InitAssemblyProperties();
+            _projectMock.Setup(p => p.UniqueName).Returns(UniqueNameDifferent);
+            _objectUnderTest = new ProjectHelper(_projectMock.Object);
+
+            TestStandardInitialState();
+            TestAssemblyPropertiesInitialized();
+            Assert.AreEqual(UniqueNameDifferent, _objectUnderTest.UniqueName, true);
+        }
+
+        public void TestInitialStateEmptyProperties()
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: done

**Comment by amanda-tarafa** [comment]
Addressed @iantalarico comments and suggestions.

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs [hunk 10]
```diff
@@ -35,78 +35,46 @@ public void BeforeEach()
...
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException), AllowDerivedTypes = true)]
+        public void TestConstructorNoProperties() => TestConstructorByFullNamePropertiesNull(ProjectFullName, null);
 
         [TestMethod]
         [ExpectedException(typeof(ArgumentException), AllowDerivedTypes = true)]
-        public void TestConstructorNoProperties()
+        public void TestConstructorNoPropertiesNoFullName() => TestConstructorByFullNamePropertiesNull(null, null);
+
+        private void TestConstructorByProjectKind(string projectKind)
```
Comment (iantalarico) [inline GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs]: I would add a comment here given a small explanation, especially given this is excepted to throw.

Same below

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs [hunk 11]
```diff
@@ -79,66 +96,41 @@ private void InitEmptyProperties()
...
-            _objectUnderTest = new ProjectHelper(_projectMock.Object);
-
-            TestStandardInitialState();
-            TestAssemblyPropertiesInitialized();
-            Assert.AreEqual(UniqueNameSeparator, _objectUnderTest.UniqueName, true);
-        }
-
-        [TestMethod]
-        public void TestInitialStateUniqueNameDifferent()
+        private void TestInitialStateAssemblyPropertiesByUniqueName(string uniqueName)
```
Comment (iantalarico) [inline GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: Suggestion (not required): You could have one function and pass in the expected value as well.

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs [hunk 12]
```diff
@@ -79,66 +96,41 @@ private void InitEmptyProperties()
...
         {
             InitAssemblyProperties();
-            _projectMock.Setup(p => p.UniqueName).Returns(UniqueNameDifferent);
+            _projectMock.Setup(p => p.UniqueName).Returns(uniqueName);
             _objectUnderTest = new ProjectHelper(_projectMock.Object);
 
-            TestStandardInitialState();
-            TestAssemblyPropertiesInitialized();
-            Assert.AreEqual(UniqueNameDifferent, _objectUnderTest.UniqueName, true);
+            AssertStandardInitialState();
```
Comment (iantalarico) [inline GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: This is only used in one place and is short.  May be worth moving into this function.

Same below

Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: Refactored the code a little bit, this is not applicable anymore.

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs (hunk 11 above)
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperUnitTest.cs]: Refactored the code a little bit, this is not applicable anymore.

File: GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs (hunk 10 above)
Comment (amanda-tarafa) [reply GoogleCloudExtension/GoogleCloudExtensionUnitTests/SolutionUtils/ProjectHelperConstructorUnitTests.cs]: Done

**Comment by amanda-tarafa** [comment]
Great, I will add more tests to around all of this in another PR and will take a look at that. Thanks!

**Comment by ILMTitan** [comment]
Fixes #869

**Pull request merged**